File: backend/models/bigquery_cost.py
"""Pydantic models for the BigQuery cost dashboard.

Fed by `INFORMATION_SCHEMA.JOBS_BY_PROJECT` aggregated by user, statement type,
referenced table and top queries. Custo estimado on-demand a partir de
`total_bytes_billed`.
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class BqCostUserRow(BaseModel):
    user_email: str
    jobs: int
    bytes_billed: int
    slot_ms: int
    cost_usd: Decimal
    cost_brl: Decimal


class BqCostStatementRow(BaseModel):
    statement_type: str
    jobs: int
    bytes_billed: int
    slot_ms: int
    cost_usd: Decimal
    cost_brl: Decimal


class BqCostTableRow(BaseModel):
    table_fqn: str
    jobs: int
    bytes_billed: int
    cost_usd: Decimal
    cost_brl: Decimal


class BqCostUserDailyPoint(BaseModel):
    """Custo de análise (query) de um usuário num dia — série temporal empilhada."""

    day: date
    user_email: str
    cost_usd: Decimal
    cost_brl: Decimal


class BqCostQueryRow(BaseModel):
    job_id: str
    user_email: str | None
    statement_type: str | None
    creation_time: str
    bytes_billed: int
    slot_ms: int
    cost_usd: Decimal
    cost_brl: Decimal
    query_preview: str
    region: str


class BqUserQueryGroup(BaseModel):
    """Padrão de query de um usuário (literais normalizados), agregado."""

    pattern_preview: str
    sample_query: str
    statement_type: str | None
    jobs: int
    bytes_billed: int
    slot_ms: int
    cost_usd: Decimal
    cost_brl: Decimal


class BqUserQueriesResponse(BaseModel):
    """Padrões de query mais caros de um usuário (drill-down da tabela por usuário)."""

    user_email: str
    from_date: date
    to_date: date
    currency: str
    exchange_rate: Decimal
    total_cost_usd: Decimal
    total_cost_brl: Decimal
    groups: list[BqUserQueryGroup]
    cached: bool = False
    fetched_at: str


class BqCostLimitStatus(BaseModel):
    """Estado do limite diário de gasto com BigQuery (escopo projeto).

    `status` resume o dia: `unset` (sem limite), `ok`, `warning` (>= warn_pct),
    `exceeded` (gasto real já passou do limite) e `projected` (gasto atual ok,
    mas a projeção linear do dia estoura).

    `enforcement` reflete o bloqueio real via quota do GCP (consumer override em
    `bigquery.googleapis.com/quota/query/usage`): `enforced` (override ativo),
    `off` (limite salvo mas sem bloqueio aplicado), `stray` (há override sem
    limite salvo), `error` (falha ao consultar/aplicar). Ao estourar, TODA query
    on-demand do projeto falha até a meia-noite do Pacífico.
    """

    day: date
    timezone: str
    regions: list[str]
    limit_brl: Decimal | None
    warn_pct: Decimal
    spent_brl: Decimal
    spent_bytes: int
    spent_usd: Decimal
    projected_brl: Decimal
    pct_used: Decimal | None
    day_fraction_elapsed: Decimal
    status: str
    enforcement: str
    enforced_limit_mib: int | None = None
    enforced_limit_tib: Decimal | None = None
    quota_effective_tib: Decimal | None = None
    enforce_error: str | None = None
    exchange_rate: Decimal
    price_usd_per_tib: Decimal
    updated_at: str | None
    updated_by: str | None
    fetched_at: str


class BqCostDashboardResponse(BaseModel):
    from_date: date
    to_date: date
    regions: list[str]
    exchange_rate: Decimal
    # Tarifa efetiva média (USD/TiB) — blended real, não mais o flat fixo.
    price_usd_per_tib: Decimal
    total_jobs: int
    total_bytes_billed: int
    total_slot_ms: int
    # Total da aba = custo total do serviço BigQuery no billing (análise + storage
    # + outros), para bater com a linha "BigQuery" da aba Google Cloud.
    total_cost_usd: Decimal
    total_cost_brl: Decimal
    by_user: list[BqCostUserRow]
    by_statement_type: list[BqCostStatementRow]
    by_table: list[BqCostTableRow]
    top_queries: list[BqCostQueryRow]
    by_user_daily: list[BqCostUserDailyPoint] = []
    cached: bool = False
    fetched_at: str
    # Moeda nativa do billing (BRL/USD) usada na calibração.
    currency: str = "USD"
    # Se o custo veio calibrado pelo billing export (True) ou do flat on-demand (False).
    calibrated: bool = False
    # Quebra do custo total do serviço BigQuery.
    analysis_cost_usd: Decimal = Decimal("0")
    analysis_cost_brl: Decimal = Decimal("0")
    storage_cost_usd: Decimal = Decimal("0")
    storage_cost_brl: Decimal = Decimal("0")
    other_cost_usd: Decimal = Decimal("0")
    other_cost_brl: Decimal = Decimal("0")
    # Tarifa efetiva por região (USD/TiB), derivada do billing.
    price_by_region: dict[str, Decimal] = {}

File: frontend/src/services/api/bigquery-cost.ts
import { fetchJsonWithTimeout } from "./http";

export type BqCostUserRow = {
  user_email: string;
  jobs: number;
  bytes_billed: number;
  slot_ms: number;
  cost_usd: string;
  cost_brl: string;
};

export type BqCostStatementRow = {
  statement_type: string;
  jobs: number;
  bytes_billed: number;
  slot_ms: number;
  cost_usd: string;
  cost_brl: string;
};

export type BqCostTableRow = {
  table_fqn: string;
  jobs: number;
  bytes_billed: number;
  cost_usd: string;
  cost_brl: string;
};

export type BqCostUserDailyPoint = {
  day: string;
  user_email: string;
  cost_usd: string;
  cost_brl: string;
};

export type BqCostQueryRow = {
  job_id: string;
  user_email: string | null;
  statement_type: string | null;
  creation_time: string;
  bytes_billed: number;
  slot_ms: number;
  cost_usd: string;
  cost_brl: string;
  query_preview: string;
  region: string;
};

export type BqCostDashboardResponse = {
  from_date: string;
  to_date: string;
  regions: string[];
  exchange_rate: string;
  price_usd_per_tib: string;
  total_jobs: number;
  total_bytes_billed: number;
  total_slot_ms: number;
  total_cost_usd: string;
  total_cost_brl: string;
  by_user: BqCostUserRow[];
  by_statement_type: BqCostStatementRow[];
  by_table: BqCostTableRow[];
  top_queries: BqCostQueryRow[];
  by_user_daily: BqCostUserDailyPoint[];
  cached: boolean;
  fetched_at: string;
  currency: string;
  calibrated: boolean;
  analysis_cost_usd: string;
  analysis_cost_brl: string;
  storage_cost_usd: string;
  storage_cost_brl: string;
  other_cost_usd: string;
  other_cost_brl: string;
  price_by_region: Record<string, string>;
};

export function fetchBigQueryCostDashboard(
  url: string,
): Promise<BqCostDashboardResponse> {
  return fetchJsonWithTimeout<BqCostDashboardResponse>(url, {
    timeoutMs: 90000,
    errorMessage: "Falha ao carregar custos do BigQuery.",
  });
}

export type BqUserQueryGroup = {
  pattern_preview: string;
  sample_query: string;
  statement_type: string | null;
  jobs: number;
  bytes_billed: number;
  slot_ms: number;
  cost_usd: string;
  cost_brl: string;
};

export type BqUserQueriesResponse = {
  user_email: string;
  from_date: string;
  to_date: string;
  currency: string;
  exchange_rate: string;
  total_cost_usd: string;
  total_cost_brl: string;
  groups: BqUserQueryGroup[];
  cached: boolean;
  fetched_at: string;
};

export function fetchBqUserQueries(
  url: string,
): Promise<BqUserQueriesResponse> {
  return fetchJsonWithTimeout<BqUserQueriesResponse>(url, {
    timeoutMs: 90000,
    errorMessage: "Falha ao carregar as queries do usuário.",
  });
}

export type BqLimitStatusValue =
  | "unset"
  | "ok"
  | "projected"
  | "warning"
  | "exceeded";

export type BqEnforcementValue = "enforced" | "off" | "stray" | "error";

export type BqCostLimitStatus = {
  day: string;
  timezone: string;
  regions: string[];
  limit_brl: string | null;
  warn_pct: string;
  spent_brl: string;
  spent_bytes: number;
  spent_usd: string;
  projected_brl: string;
  pct_used: string | null;
  day_fraction_elapsed: string;
  status: BqLimitStatusValue;
  enforcement: BqEnforcementValue;
  enforced_limit_mib: number | null;
  enforced_limit_tib: string | null;
  quota_effective_tib: string | null;
  enforce_error: string | null;
  exchange_rate: string;
  price_usd_per_tib: string;
  updated_at: string | null;
  updated_by: string | null;
  fetched_at: string;
};

function limitUrl(apiBase: string): string {
  return `${apiBase.replace(/\/$/, "")}/api/bigquery-cost/limit`;
}

export function fetchBqLimitStatus(url: string): Promise<BqCostLimitStatus> {
  return fetchJsonWithTimeout<BqCostLimitStatus>(url, {
    timeoutMs: 60000,
    errorMessage: "Falha ao carregar o limite do BigQuery.",
  });
}

export function putBqLimit(
  apiBase: string,
  body: { limit_brl: number; warn_pct?: number },
  userEmail?: string,
): Promise<BqCostLimitStatus> {
  return fetchJsonWithTimeout<BqCostLimitStatus>(limitUrl(apiBase), {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      ...(userEmail ? { "X-User-Email": userEmail } : {}),
    },
    body: JSON.stringify(body),
    timeoutMs: 60000,
    errorMessage: "Falha ao salvar o limite.",
  });
}

export function deleteBqLimit(
  apiBase: string,
  userEmail?: string,
): Promise<BqCostLimitStatus> {
  return fetchJsonWithTimeout<BqCostLimitStatus>(limitUrl(apiBase), {
    method: "DELETE",
    headers: userEmail ? { "X-User-Email": userEmail } : undefined,
    timeoutMs: 60000,
    errorMessage: "Falha ao remover o limite.",
  });
}
